soundwire: send multi sections in one BPT stream

bardliao opened the pull request:
Currently we send a BRA message with a start address with continuous registers in a BPT stream. However, a codec may need to write different register sections shortly. It makes more sense to send different register sections in a BPT stream and no need to close/open the BPT stream repeatedly.

bardliao commented on drivers/soundwire/cadence_master.c:
```diff
@@ -2676,6 +2702,18 @@ int sdw_cdns_check_read_response(struct device *dev, u8 *dma_buffer, int dma_buf
 		counter++;
 		counter &= GENMASK(3, 0);
+
+		if (buffer_size == total_num_bytes && (i + 1) < num_frames) {
+			sec_index++;
+			if (sec_index >= num_sec) {
+				dev_err(dev, "%s: incorrect section index %d i %d\n",
+					__func__, sec_index, i);
+				return -EINVAL;
+			}
+			p_buf = sec[sec_index].buf;
+			buffer_size = sec[sec_index].len;
+			total_num_bytes = 0;
+		}
```
`buffer_size` is the buffer size of each section and `total_num_bytes` is the buffer size that has been written. They should be exactly the same when a section is sent. Then reset `buffer_size` and `total_num_bytes` for the next section.

shumingfan reviewed (dismissed):
Tested-by: Shuming Fan <<shumingf@realtek.com>>

bardliao committed:
soundwire: introduce BPT section

Currently we send a BRA message with a start address with continuous
registers in a BPT stream. However, a codec may need to write different
register sections shortly. Introduce a register section in struct
sdw_btp_msg which contain register start address, length, and buffer.
This commit uses only 1 section for each BPT message. And we need to add
up all BPT section lengh and check if it reach maximum BPT bytes.
No function changes.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
soundwire: pass sdw_bpt_section to cdns BPT helpers

We can get start_register, data_size, and buffer data from the new
sdw_bpt_section parameter. Also, handle all register sections in the
cdns BRA helpers. No function changes as section number is 1.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

bardliao committed:
soundwire: intel_ace2x: handle multi BPT sections

Calculate required PDI buffer and pass the section number to the cdns
BPT helpers.

Signed-off-by: Bard Liao <yung-chuan.liao@linux.intel.com>

ranj063 commented on drivers/soundwire/bus.h:
```diff
@@ -73,21 +73,31 @@ struct sdw_msg {
 
 /**
- * struct sdw_btp_msg - Message structure
+ * struct sdw_btp_section - Message structure
```
Message section structure?

ranj063 commented on drivers/soundwire/bus.h:
```diff
@@ -73,21 +73,31 @@ struct sdw_msg {
  * but a practical limit of SDW_BPT_MSG_MAX_BYTES is enforced.
- * @dev_num: Slave device number
- * @flags: transfer flags, indicate if xfer is read or write
  * @buf: message data buffer (filled by host for write, filled
```
section data buffer?

ranj063 commented on drivers/soundwire/intel_ace2x.c:
```diff
@@ -369,7 +367,7 @@ static int intel_ace2x_bpt_wait(struct sdw_intel *sdw, struct sdw_slave *slave,
 		ret = sdw_cdns_check_read_response(cdns->dev, sdw->bpt_ctx.dmab_rx_bdl.area,
 						   sdw->bpt_ctx.pdi1_buffer_size,
-						   msg->sec[0].buf, msg->sec[0].len, sdw->bpt_ctx.num_frames,
+						   msg->sec, 1, sdw->bpt_ctx.num_frames,
```
you set msg->sections to 1 in the first patch isnt it? Why not use that here and the above 2 places?

bardliao replied:
`msg->sections = 1` was set in debugfs which provides an example to use the helpers. And intel_ace2x_bpt_open_stream() didn't find the buffer size for other sections other than the first section yet. No matter what `msg->sections' is set, it can handle 1 section. That's why I pass `1` here and will change to `msg->sections' when it can handle other sections in the next commit.

ranj063 commented:
> > @bardliao this PR is not easy to review. Can you break it down by splitting into a couple of different patches and incrementally add the sections. For example, introduce the concept of sections first, then modify the current code so that you have only one section and then finally add the logic for multiple sections
> 
> @ranj063 Updated.

Thanks @bardliao. A couple of minor comments only now
